File: src/thread_pool.c
#include "thread_pool.h"

#include <assert.h>
#include <pthread.h>
#include <stdbool.h>
#include <stdlib.h>

struct task {
    work_function_t function;
    void *aux;
};

struct thread_pool {
    pthread_t *threads;
    size_t num_threads;

    queue_t *work_queue;
    pthread_cond_t work_in_queue_cond;

    bool in_use;
};

void *get_work_from_queue(thread_pool_t *thread_pool) {
    task_t *task = NULL;
    for (task = (task_t *) queue_dequeue(thread_pool->work_queue); task != NULL;
         task = (task_t *) queue_dequeue(thread_pool->work_queue)) {
        assert(task != NULL);
        assert(task->function != NULL);
        task->function(task->aux);
        free(task);
    }
    return NULL;
}

/**
 * Creates a new heap-allocated thread pool with the given number of worker threads.
 * All worker threads should start immediately so they can perform work
 * as soon as thread_pool_add_work() is called.
 *
 * @param num_worker_threads the number of threads in the pool
 * @return a pointer to the new thread pool
 */
thread_pool_t *thread_pool_init(size_t num_worker_threads) {
    thread_pool_t *thread_pool = (thread_pool_t *) calloc(1, sizeof(thread_pool_t));
    if (thread_pool == NULL) {
        return NULL;
    }

    thread_pool->num_threads = num_worker_threads;
    thread_pool->work_queue = queue_init();

    thread_pool->threads =
        (pthread_t *) calloc(thread_pool->num_threads, sizeof(pthread_t));
    if (thread_pool->threads == NULL) {
        return NULL;
    }

    for (size_t i = 0; i < thread_pool->num_threads; i++) {
        // queue_enqueue(thread_pool->work_queue, NULL);
        int pthread_error = pthread_create(&thread_pool->threads[i], NULL,
                                           get_work_from_queue, thread_pool);
        assert(pthread_error == 0);
    }

    pthread_cond_init(&thread_pool->work_in_queue_cond, NULL);

    thread_pool->in_use = true;

    return thread_pool;
}

/**
 * Adds work to a thread pool.
 * The work will be performed by a worker thread as soon as all previous work is finished.
 *
 * @param pool the thread pool to perform the work
 * @param function the function to call on a thread in the thread pool
 * @param aux the argument to call the work function with
 */
void thread_pool_add_work(thread_pool_t *pool, work_function_t function, void *aux) {
    task_t *task = calloc(1, sizeof(task_t));
    assert(task != NULL);
    task->function = function;
    task->aux = aux;
    assert(function == 0x4ca930);
    queue_enqueue(pool->work_queue, task);
}

/**
 * Waits for all work added to a thread pool to finish,
 * then frees all resources associated with a heap-allocated thread pool.
 * A special value (e.g. NULL) can be put in the work queue to mark the end of the work.
 * thread_pool_add_work() cannot be used on this pool once this function is called.
 *
 * @param pool the thread pool to close
 */
void thread_pool_finish(thread_pool_t *pool) {
    for (size_t i = 0; i < pool->num_threads; i++) {
        queue_enqueue(pool->work_queue, NULL);
    }
    for (size_t i = 0; i < pool->num_threads; i++) {
        pthread_join(pool->threads[i], NULL);
    }

    free(pool->threads);

    queue_free(pool->work_queue);

    free(pool);
}

File: src/thread_pool.rs
use std::sync::{mpsc, Arc, Mutex};
use std::thread;

// https://doc.rust-lang.org/book/ch20-02-multithreaded.html
pub struct ThreadPool<'scope, 'env> {
    pub workers: Vec<Worker<'scope>>,
    sender: mpsc::Sender<Job<'env>>,
}

// impl<'scope, 'env> Clone for ThreadPool<'scope, 'env> {
//     fn clone(&self) -> Self {
//         Self {
//             workers: self.workers.clone(),
//             sender: self.sender.clone(),
//         }
//     }
// }

type Job<'env> = Box<dyn FnOnce() + Send + 'env>;

impl<'scope, 'env> ThreadPool<'scope, 'env> {
    /* *
     * Creates a new thread pool with the given number of worker threads.
     * All worker threads should start immediately so they can perform work
     * as soon as self.execute() is called.
     *
     * @param num_worker_threads the number of threads in the pool
     * @param scope the thread pool's variable lifetime scope.
     * @return the new thread pool
     * Lifetimes 'scope and 'env are documented here
     * https://doc.rust-lang.org/nightly/std/thread/fn.scope.html#lifetimes
     */
    pub fn new(
        num_worker_threads: usize,
        scope: &'scope thread::Scope<'scope, 'env>,
    ) -> ThreadPool<'scope, 'env> {
        assert!(num_worker_threads > 0);

        let (sender, receiver) = mpsc::channel();
        let receiver = Arc::new(Mutex::new(receiver));

        let mut workers = Vec::with_capacity(num_worker_threads);

        for id in 0..num_worker_threads {
            workers.push(Worker::new(id, scope, Arc::clone(&receiver)));
        }

        ThreadPool { workers, sender }
    }
    /* *
     * Adds work to a thread pool.
     * The work will be performed by a worker thread as soon as all previous work is finished.
     *
     * @param f the function to call on a thread in the thread pool
     * Note the lifetime 'env on function f in accordance to
     * https://doc.rust-lang.org/nightly/std/thread/fn.scope.html#lifetimes
     */
    pub fn execute<F>(&self, f: F)
    where
        F: FnOnce() + Send + 'env,
    {
        let job = Box::new(f);

        self.sender.send(job).unwrap();
    }
}

pub struct Worker<'scope> {
    pub id: usize,
    pub thread: thread::ScopedJoinHandle<'scope, ()>,
}

impl<'scope, 'env> Worker<'scope> {
    /* *
     * Worker that processes jobs added to the threadpool as they are added.
     *
     * @param id the worker (thread's) identifying number.
     * @param scope the variable lifetime scope to use.
     * @param receiver is the queue that is waited on for new jobs to execute.
     * @return the the worker
     */
    fn new(
        id: usize,
        scope: &'scope thread::Scope<'scope, 'env>,
        receiver: Arc<Mutex<mpsc::Receiver<Job<'env>>>>,
    ) -> Worker<'scope> {
        let thread = scope.spawn(move || loop {
            let job = {
                let lock = receiver
                    .lock()
                    .expect("Worker thread unable to lock the job receiver");
                lock.recv()
            };

            // println!("Worker {} got a job; executing.", id);

            let job = match job {
                Ok(job) => job,
                // ThreadPool has been dropped
                Err(..) => break,
            };

            job();

            // println!("Worker {} completed its job; executing.", id);
        });

        Worker { id, thread }
    }
}
